crates/jstzd/src/task/jstzd.rs

```diff
@@ -0,0 +1,117 @@
+use super::{octez_node::OctezNode, Task};
+use anyhow::Result;
+use async_dropper_simple::AsyncDrop;
+use async_trait::async_trait;
+use octez::r#async::node_config::OctezNodeConfig;
+use tokio::task::JoinHandle;
+
+pub struct Jstzd {
+    octez_node: OctezNode,
+}
+
+#[derive(Clone)]
+pub struct JstzdConfig {
+    octez_node_config: OctezNodeConfig,
+}
+
+impl JstzdConfig {
+    pub fn new(octez_node_config: OctezNodeConfig) -> Self {
+        Self { octez_node_config }
+    }
+}
+
+#[async_trait]
+impl Task for Jstzd {
+    type Config = JstzdConfig;
+
+    async fn spawn(config: Self::Config) -> Result<Self> {
+        Ok(Self {
+            octez_node: OctezNode::spawn(config.octez_node_config.clone())
+                .await
+                .unwrap(),
+        })
+    }
+
+    async fn kill(&mut self) -> Result<()> {
+        self.octez_node.kill().await
+    }
+
+    async fn health_check(&self) -> Result<bool> {
+        self.octez_node.health_check().await
+    }
+}
+
+pub struct JstzdServer {
+    jstzd: Option<Jstzd>,
+    jstzd_config: JstzdConfig,
+    jstzd_server_port: u16,
+    server_handle: Option<JoinHandle<()>>,
+}
+
+#[async_trait]
+impl AsyncDrop for JstzdServer {
+    async fn async_drop(&mut self) {
+        if let Err(e) = self.stop().await {
+            eprintln!("failed to stop jstzd server: {:?}", e);
+        }
+    }
+}
+
+impl JstzdServer {
+    pub fn new(config: JstzdConfig, port: u16) -> Self {
+        Self {
+            jstzd: None,
+            jstzd_config: config,
+            jstzd_server_port: port,
+            server_handle: None,
+        }
+    }
+
+    pub async fn health_check(&self) -> bool {
+        let res =
+            reqwest::get(format!("http://localhost:{}", self.jstzd_server_port)).await;
+        if res.is_err() {
+            return false;
+        }
+
+        if let Some(jstzd) = &self.jstzd {
+            if let Ok(v) = jstzd.health_check().await {
+                return v;
+            }
+        }
+
+        false
+    }
+
+    pub async fn stop(&mut self) -> Result<()> {
+        let mut err = None;
+        if let Some(mut jstzd) = self.jstzd.take() {
+            if let Err(e) = jstzd.kill().await {
+                err.replace(e);
+            };
+        }
+        if let Some(server) = self.server_handle.take() {
+            server.abort();
+        }
+        match err {
+            Some(e) => Err(e),
+            None => Ok(()),
+        }
+    }
+
+    pub async fn run(&mut self) -> Result<()> {
+        self.jstzd
+            .replace(Jstzd::spawn(self.jstzd_config.clone()).await?);
+
+        let router =
+            axum::Router::new().route("/", axum::routing::get(http::StatusCode::OK));
+        let listener =
+            tokio::net::TcpListener::bind(("0.0.0.0", self.jstzd_server_port)).await?;
+
+        let handle = tokio::spawn(async {
+            axum::serve(listener, router).await.unwrap();
+        });
+        self.server_handle.replace(handle);
```

Comment:
don't we need to await on the handle here? otherwise the server will shutdown if the main fnction exits.

Reply:
No. According to the [doc](https://docs.rs/tokio/latest/tokio/task/fn.spawn.html),

> The provided future will start running in the background immediately when spawn is called, even if you don’t await the returned JoinHandle.

If this handle is awaited, this method only returns when the server is shutdown. Testing is all included in the new integration test.

Reply:
i understand the server starts correctly based on the integration test. i was wondering say if we want to spin up the server like the following code:
```
#[tokio]
fn main () {
  // this will return immediately and main would exit as we don't await on the join handler
  JstzdServer::run().await;
}
```
then the spawned jstzd would be shutdown as the main function would exit immeidiately?
The doc says:

> There is no guarantee that a spawned task will execute to completion. When a runtime is shutdown, all outstanding tasks are dropped, regardless of the lifecycle of that task.

Reply:
Yes, then the question is how this jstzd server will be launched and this is totally up to discussions. If we just want to run the server like how most servers are run, we just need call `axum::serve` and don't return the handle at all, which will work with the code snippet. However, from what I see now in the issues defined so far, it seems that in jstzd main there will be a readiness check after jstzd server is launched, kind of like the following based on what I understand:

* start jstzd server
* wait until all the components are up and running
  * ok: inform users that they are good to go and then keep everything running until users terminate the sandbox
  * timeout: shut down everything and show errors

Then in this case we won't run the server as what you described but somehow keep the join handle alive.

Reply:
> ok: inform users that they are good to go and then keep everything running until users terminate the sandbox

In this case, don't we still need a way to await on the join handler? maybe we can add a method that waits for the join handler?

Reply:
Not really. As long as the join handler does not get dropped, the task keeps running. I imagine that jstzd main would be like
```rs
let handle = jstzd.run(...)
tokio::select!{
  // listen to signals and cancel the handle or call stop on request
}
```

If we change the implementation so that it's like `jstzd_server.run().await;`, it should also work. It's just that in main there will be a `tokio::spawn` if we want to do something else after `server.run().await`. I'm not sure about this because then the server instance will be moved into spawn and I'm not sure if rust will complain when we try to do call that instance.

Reply:
I initially scoped this function to do the run checks and wait for sigs but splitting the server orchestration and web server launch makes sense. Could you create a linear issue to track the server orchestration work please?

Reply:
https://linear.app/tezos/issue/JSTZ-189